### PR TITLE
feat: Persist chargeprices

### PR DESCRIPTION
**Pull request by x-platformcoder**
## Description
Persisting charge prices from messages with business reasoncode "D08" 

[x] Create seperate eventhandler for chargeprices
[x] Persist chargeprices from chargeprice update

## References

* #1292

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandler.cs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        private async Task AcceptValidOperationsAsync(
+            IReadOnlyCollection<ChargeOperationDto> operationsToBeConfirmed,
+            DocumentDto document)
+        {
+            if (operationsToBeConfirmed.Any())
+            {
+                await _chargeCommandReceiptService.AcceptAsync(
+                    new ChargeCommand(document, operationsToBeConfirmed)).ConfigureAwait(false);
+            }
+        }
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandler.cs:122]: This part is duplication from `ChargeEventHandler` (formerly known as `ChargeCommandReceivedEventHandler`). I suggest we introduce a new class with these methods.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/Charge.cs [hunk 2]
```diff
@@ -147,6 +147,20 @@ public void CancelStop(ChargePeriod chargePeriod)
             _periods.Add(chargePeriod);
         }
 
+        public void UpdatePrices(Instant startDate, Instant? endDate, IReadOnlyList<Point> newPrices)
+        {
+            ArgumentNullException.ThrowIfNull(endDate);
+            ArgumentNullException.ThrowIfNull(newPrices);
+            if (newPrices.Count == 0) return;
+            var removePoints = _points.Where(x => x.Time >= startDate && x.Time <= endDate).ToList();
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/Charge.cs:155]: I suggest we put this statement in a method with a name that makes it simple to understand why we want to remove these points.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Charges/ChargeCommandReceiverEndpoint.cs [hunk 3]
```diff
@@ -24,14 +25,14 @@ namespace GreenEnergyHub.Charges.FunctionHost.Charges
     public class ChargeCommandReceiverEndpoint
     {
         public const string FunctionName = nameof(ChargeCommandReceiverEndpoint);
-        private readonly IChargeHandler _chargeHandler;
+        private readonly IChargeCommandReceivedEventHandler _chargeCommandReceivedEventCommandReceivedEventHandler;
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Charges/ChargeCommandReceiverEndpoint.cs:28]: That's a `longNameLongName` ;-)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Charges/Handlers/ChargePricesHandlerTests.cs [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+            receiptService
+                .Setup(s => s.AcceptAsync(It.IsAny<ChargeCommand>()))
+                .Callback<ChargeCommand>(_ => confirmed = true);
+
+            // Act
+            await sut.HandleAsync(receivedEvent);
+
+            // Assert
+            unitOfWork.Verify(x => x.SaveChangesAsync(), Times.AtLeastOnce());
+            Assert.True(confirmed);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Charges/Handlers/ChargePricesHandlerTests.cs:73]:
```suggestion
            confirmed.Should().Be(true);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Charges/Handlers/ChargePricesHandlerTests.cs [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+                .ReturnsAsync(charge);
+
+            receiptService.Setup(s => s.RejectAsync(It.IsAny<ChargeCommand>(), It.IsAny<ValidationResult>()))
+                .Callback<ChargeCommand, ValidationResult>((_, _) => rejected = true);
+
+            // Act
+            await sut.HandleAsync(receivedEvent);
+
+            // Assert
+            Assert.True(rejected);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Charges/Handlers/ChargePricesHandlerTests.cs:103]:
```suggestion
            rejected.Should().Be(true);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Charges/Handlers/ChargePricesHandlerTests.cs [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+            ChargePriceEventHandler sut)
+        {
+            // Arrange
+            var validationResult = GetFailedValidationResult();
+            SetupValidators(inputValidator, businessValidator, validationResult);
+            var rejected = false;
+            var charge = chargeBuilder.Build();
+            chargeRepository
+                .Setup(r => r.GetOrNullAsync(It.IsAny<ChargeIdentifier>()))
+                .ReturnsAsync(charge);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Charges/Handlers/ChargePricesHandlerTests.cs:94]:
```suggestion
            SetupRepositories(chargeRepository, charge);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/GreenEnergyHub.Charges.Tests.csproj [hunk 7]
```diff
@@ -32,6 +32,7 @@ limitations under the License.
       <ProjectReference Include="..\GreenEnergyHub.Charges.Domain\GreenEnergyHub.Charges.Domain.csproj" />
       <ProjectReference Include="..\GreenEnergyHub.Charges.FunctionHost\GreenEnergyHub.Charges.FunctionHost.csproj" />
       <ProjectReference Include="..\GreenEnergyHub.Charges.Infrastructure\GreenEnergyHub.Charges.Infrastructure.csproj" />
+      <ProjectReference Include="..\GreenEnergyHub.Charges.IntegrationTest.Core\GreenEnergyHub.Charges.IntegrationTest.Core.csproj" />
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/GreenEnergyHub.Charges.Tests.csproj:35]: We should not reference any integrationtest-projects in the Tests project, and I also don't see any code relying on this reference.
```suggestion
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/GreenEnergyHub.Charges.Application.csproj [hunk 8]
```diff
@@ -27,6 +27,7 @@ limitations under the License.
   <ItemGroup>
     <ProjectReference Include="..\GreenEnergyHub.Charges.Core\GreenEnergyHub.Charges.Core.csproj" />
     <ProjectReference Include="..\GreenEnergyHub.Charges.Domain\GreenEnergyHub.Charges.Domain.csproj" />
+    <ProjectReference Include="..\GreenEnergyHub.Charges.TestCore\GreenEnergyHub.Charges.TestCore.csproj" />
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/GreenEnergyHub.Charges.Application.csproj:30]: We should not reference any test-projects in production code, and I also don't see any code relying on this reference.
```suggestion
```

**Review by HenrikSommer — COMMENTED** [review]
Nice work, I've left a few comments and suggestions.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandler.cs (hunk 1 above)
Comment (x-platformcoder) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandler.cs:122]: added the 2 methods to the chargecommandreceiptsservice class

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandler.cs [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+            var operations = commandReceivedEvent.Command.ChargeOperations.ToArray();
+
+            for (var i = 0; i < operations.Length; i++)
+            {
+                var operation = operations[i];
+                var charge = await GetChargeAsync(operation).ConfigureAwait(false);
+                if (charge is null)
+                {
+                    throw new InvalidOperationException($"Chargeinformation ID '{operation.ChargeId}' does not exist.");
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandler.cs:71]:
```suggestion
                    throw new InvalidOperationException($"Charge ID '{operation.ChargeId}' does not exist.");
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/Charge.cs [hunk 10]
```diff
@@ -147,6 +147,26 @@ public void CancelStop(ChargePeriod chargePeriod)
             _periods.Add(chargePeriod);
         }
 
+        public void UpdatePrices(Instant startDate, Instant? endDate, IReadOnlyList<Point> newPrices)
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/Charge.cs:150]: I'm a bit worried about this, as it uses the `ChargeOperationDto's` `StartDateTime` and `EndDateTime`. 

Note `ChargePeriodFactory.CreateFromChargeOperationDto()` will do this: `chargeOperationDto.EndDateTime.TimeOrEndDefault());` so I believe with the current implementation we will remove move points than intended by the price owner.

Instead, I think we should somehow carry the below `<timeInterval>` as provided by the price owner, which tells us exactly the start and end of the price series.

````
<cim:timeInterval>
 <cim:start>2022-07-31T22:00Z</cim:start>
 <cim:end>2022-08-31T22:00Z</cim:end>
</cim:timeInterval>
````

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/Charge.cs [hunk 11]
```diff
@@ -147,6 +147,26 @@ public void CancelStop(ChargePeriod chargePeriod)
...
+            RemoveExistingChargePrices(startDate, endDate);
+            _points.AddRange(newPrices);
+        }
+
+        private void RemoveExistingChargePrices(
+            Instant startDate,
+            Instant? endDate)
+        {
+            ArgumentNullException.ThrowIfNull(endDate);
+            var removePoints = _points.Where(x => x.Time >= startDate && x.Time <= endDate).ToList();
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/Charge.cs:163]: The latest point that needs to be removed, will have a `Time` that lies before the `endDate`, hence the `=` needs to be removed.

```suggestion
            var removePoints = _points.Where(x => x.Time >= startDate && x.Time < endDate).ToList();
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/IChargeRepository.cs [hunk 12]
```diff
@@ -15,6 +15,7 @@
 using System;
 using System.Collections.Generic;
 using System.Threading.Tasks;
+using NodaTime;
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/IChargeRepository.cs:18]: Can be removed

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/IChargeEventHandler.cs [hunk 13]
```diff
@@ -17,15 +17,8 @@
 
 namespace GreenEnergyHub.Charges.Application.Charges.Handlers
 {
-    /// <summary>
-    /// Delegates a charge depending on if it contains
-    /// Charge Prices or Charge information.
-    /// </summary>
-    public interface IChargeHandler
+    public interface IChargeEventHandler
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/IChargeEventHandler.cs:20]: Perhaps?

```suggestion
    public interface IChargeInformationEventHandler
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/ChargeRepository.cs [hunk 14]
```diff
@@ -19,6 +19,7 @@
 using System.Threading.Tasks;
 using GreenEnergyHub.Charges.Domain.Charges;
 using Microsoft.EntityFrameworkCore;
+using NodaTime;
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Persistence/Repositories/ChargeRepository.cs:22]: Can be removed

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Configuration/ChargeCommandReceiverConfiguration.cs [hunk 15]
```diff
@@ -54,7 +54,10 @@ internal static void ConfigureServices(IServiceCollection serviceCollection)
             serviceCollection.AddScoped<ICimValidationErrorCodeFactory, CimValidationErrorCodeFactory>();
             serviceCollection.AddScoped<IAvailableChargeReceiptValidationErrorFactory,
                 AvailableChargeReceiptValidationErrorFactory>();
-            serviceCollection.AddScoped<IChargeHandler, ChargeHandler>();
+            serviceCollection.AddScoped<IChargeCommandReceivedEventHandler, ChargeCommandReceivedEventHandler>();
+            serviceCollection.AddScoped<IChargeEventHandler, ChargeEventHandler>();
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Configuration/ChargeCommandReceiverConfiguration.cs:58]: Is already added in line 47

**Review by prtandrup — COMMENTED** [review]
A few comments. Haven't looked at the tests yet

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeCommandReceiptsForOperations.cs [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+using GreenEnergyHub.Charges.Application.Charges.Acknowledgement;
+using GreenEnergyHub.Charges.Domain.Dtos.ChargeCommands;
+using GreenEnergyHub.Charges.Domain.Dtos.SharedDtos;
+using GreenEnergyHub.Charges.Domain.Dtos.Validation;
+
+namespace GreenEnergyHub.Charges.Application.Charges.Handlers
+{
+    public class ChargeCommandReceiptsForOperations : IChargeCommandReceiptsForOperations
```
[the remark this answers is not in the document]
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeCommandReceiptsForOperations.cs:25]: Ok, let's not do that now then.

[the remark this answers is not in the document]
Comment (x-platformcoder) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeCommandReceiptsForOperations.cs:25]: I found a solution after a good nights sleep.

**Review by HenrikSommer — COMMENTED** [review]
A couple of new interfaces need documentation, as proposed by SonarCloud.

**Review by Mech0z — APPROVED** [review]

**Pull request merged**